Protect claims from player-caused knockback (wind charges, spears)

FreeMonoid opened the pull request:
Adds protection against players using knockback mechanics to push entities out of claims where they lack trust. This covers wind charges, spears, and any other player-caused knockback that might be added in the future.

PvP interactions:
- Self-boosting always allowed (e.g., wind charge mobility)
- Admins bypass all checks if ignoring claims
- Container trust allows pushing other players in the claim
- In PvE worlds, knockback between players is blocked
- In PvP worlds, knockback is allowed unless:
  - Either player is a fresh spawn, or
  - Either player is in a PvP safe zone

PvE interactions:
- Armor stands and hanging entities require build trust
- Animals require container trust (if creature protection is enabled)
- Monsters are never protected

Closes #2536
Closes #2547

FreeMonoid committed:
Handle wind charge in claims in a more intuitive manner

PvP interactions:
- Self-boosting always allowed
- Admins bypass all checks if ignoring claims
- If player has container trust in a claim, they can push around other players in the claim
- Otherwise, if PvP is disabled don't allow pushing around other players
- If PvP IS enabled, allow knocking other players around
  - Unless they are fresh spawns, or
  - Either shooter or defender is in a PvP safe zone

PvE interactions:
- Armor stands and hanging entities require build trust
- Animals require container trust (if animal protection is enabled)

FreeMonoid commented:
@RoboMWM and maybe @Jikoo / @QarthO — would appreciate your input for the direction of this PR
 
The changes technically work in the current state, but on Paper we get deprecation notice:
```
[21:27:47 WARN]: [GriefPrevention] "GriefPrevention v16.18.5-7-g9c5c211" has registered a listener for org.bukkit.event.entity.EntityKnockbackByEntityEvent on method "public void me.ryanhamshire.GriefPrevention.EntityDamageHandler.onEntityKnockbackByEntity(org.bukkit.event.entity.EntityKnockbackByEntityEvent)", but the event is Deprecated. "Server performance will be affected"; please notify the authors [].
```
since https://jd.papermc.io/paper/1.21.10/org/bukkit/event/entity/EntityKnockbackByEntityEvent.html is deprecated and slated for removal

I don't know what our stance on Paper-specific additions currently are: are we OK with them at all? Should I just use reflection to decide which event we want to listen to depending on runtime platform or do we have another approach somewhere?

Jikoo commented:
Didn't realize this was legacy at first and wrote a whole long thing about how we really need to modularize the project for future support of both Paper and Spigot. Whoopsies.

The biggest problem imo is that we need a Paper dependency for the Paper events, so we need a Paper module. The main Spigot side can then just conditionally choose to use our Paper version or the regular version when registering events.

Reflective platform determination sounds good. I believe usually one targets config classes, but if we have specific listeners we want to enable, we could target the event class each is supposed to listen to for extra security.

FreeMonoid committed:
Use Paper-specific EntityKnockbackByEntityEvent when available

FreeMonoid commented:
@RoboMWM @Jikoo alright, finally got the time to take a stab at this 😅  Would love your reviews when you have time, guys!

FreeMonoid committed:
Make luredByPlayer key completely static

FreeMonoid commented on src/main/java/me/ryanhamshire/GriefPrevention/EntityDamageHandler.java:
```diff
@@ -75,13 +75,17 @@ public class EntityDamageHandler implements Listener
         this.dataStore = dataStore;
         instance = plugin;
-        luredByPlayer = new NamespacedKey(plugin, "lured_by_player");
+        // Static field initialization - safe because only one instance is created at startup
```
Yes, good point, that's a good idea! Converted it in ed57669

RoboMWM commented:
I'm still thinking about this - I know this is for legacy but especially for master and moving forward I'm still thinking about how spigot vs. paper registration should happen; if it'd be more desirable to offload specifically that to a separate class and have the actual handling elsewhere.

tho I guess this probably enters what jikoo wants to do with modules perhaps.

just haven't had the contiguous time to dedicate to how the approach should look.

FreeMonoid commented:
Yeah, makes sense, that's why I haven't even seriously thought about porting this to master: these changes are just enough to get by in the meantime 😅

FreeMonoid committed:
Abstract platform detection a bit more

FreeMonoid commented:
@Jikoo after chatting with @RoboMWM on Discord, I've added a slim abstraction (or at least what I hope you'll consider slim!) for impl-specific listeners. Should give us a bit more structure for adding new ones in the future.

Open to bikeshedding or any other thoughts!

Jikoo reviewed (approved):
This looks clean, it seems like a very neat and expandable way to handle the issue without major structural changes.

FreeMonoid committed:
Generalize knockback protection to handle spears attacks

Removes the cause filter so all player-caused knockback is subject to claim trust checks. This fixes spear attacks
bypassing claim protection due to using ENTITY_ATTACK cause instead of EXPLOSION.

FreeMonoid committed:
Handle mace smash AoE knockback on Paper

FreeMonoid commented:
@Jikoo didn't realize mace AoE smash attack actually triggered yet another type of event, `EntityPushedByEntityAttackEvent` (Paper-only, Spigot doesn't have an equivalent) — added handling for it to the Paper handler 😅 

There's a caveat though: mace with Wind Burst bypasses cancelation and causes knockback/push anyway (https://github.com/PaperMC/Paper/issues/13079)

FreeMonoid committed:
Do not mention swords in the comments, they aren't really applicable

FreeMonoid committed:
Make KnockbackProtectionHandler.java follow editorconfig standard

RoboMWM committed:
Merge branch 'legacy/v16' into wind-charge-knockback-legacy

FreeMonoid committed:
Check for `preventTheft` instead of `protectCreatures`

FreeMonoid committed:
Require access trust for pushing players, not container trust

RoboMWM committed:
Merge branch 'legacy/v16' into wind-charge-knockback-legacy
